### PR TITLE
RavenDB-22873 - Improve the deployment of multiple indexes

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22873/Improve-the-deployment-of-multiple-indexes

### Additional description

The endpoint that executes `IndexCreation.CreateIndexes` currently deploys indexes one by one.
Deploying indexes one by one generates a separate task for each index that executes the HandleDatabaseRecordChanged. This results in significant pressure on the cluster.
The optimization is achieved by using the approach applied in Smuggler, where indexes are deployed in batches of 50.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/ServerWide/Commands/Indexes/PutIndexesCommand.cs [hunk 1]
```diff
@@ -40,19 +43,48 @@ public PutIndexesCommand(string databaseName, string source, DateTime createdAt,
...
+
                 foreach (var definition in Static)
+                {
+                    if (record.Indexes.TryGetValue(definition.Name, out _) == false)
+                    {
+                        // this is not an update to an existing index. we'll check for:
+                        // - directory name collisions
+                        // - index name case sensitivity
+
+                        safeFileSystemIndexNames ??= record.Indexes.Select(x => IndexDefinitionBaseServerSide.GetIndexNameSafeForFileSystem(x.Value.Name)).ToHashSet(StringComparer.OrdinalIgnoreCase);
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Indexes/PutIndexesCommand.cs]: where this change is coming from? it looks also as a duplicated code, no test as well

Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/Commands/Indexes/PutIndexesCommand.cs]: Actually there were 3 tests that failed.
`CreateFullAndIncrementalBackupWithIndexInTheMiddle`
`CreateIndexWithTheSameLowercasedNameShouldThrow`
`CannotPutIndexWithNameClash`

That is the reason that this change was introduced.

Comment (ppekrol) [reply src/Raven.Server/ServerWide/Commands/Indexes/PutIndexesCommand.cs]: ok, but still you need to centrailize this, now you have 2 usages, one in AbstractIndexCreateController and 2nd one here.

Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/Commands/Indexes/PutIndexesCommand.cs]: done

File: src/Raven.Server/ServerWide/Commands/Indexes/PutIndexCommand.cs [hunk 2]
```diff
@@ -38,15 +40,11 @@ public PutIndexCommand(IndexDefinition definition, string databaseName, string s
...
+            var indexValidator = new StaticIndexNameValidator(record);
+            indexValidator.Validate(Definition);
+
             try
             {
-                var indexNames = record.Indexes.Select(x => x.Value.Name).ToHashSet(StringComparer.OrdinalIgnoreCase);
-
-                if (indexNames.Add(Definition.Name) == false && record.Indexes.TryGetValue(Definition.Name, out _) == false)
-                {
-                    throw new InvalidOperationException($"Can not add index: {Definition.Name} because an index with the same name but different casing already exist");
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Indexes/PutIndexCommand.cs:47]: you will throw different ex now

File: src/Raven.Server/Documents/Indexes/AbstractIndexCreateController.cs [hunk 3]
```diff
@@ -57,16 +57,6 @@ protected virtual async ValueTask ValidateStaticIndexAsync(IndexDefinition defin
...
         ServerStore.LicenseManager.AssertCanAddAdditionalAssembliesFromNuGet(definition);
 
-        var safeFileSystemIndexName = IndexDefinitionBaseServerSide.GetIndexNameSafeForFileSystem(definition.Name);
-
-        var indexWithFileSystemNameCollision = GetIndexNames().FirstOrDefault(x =>
-            x.Equals(definition.Name, StringComparison.OrdinalIgnoreCase) == false &&
-            safeFileSystemIndexName.Equals(IndexDefinitionBaseServerSide.GetIndexNameSafeForFileSystem(x), StringComparison.OrdinalIgnoreCase));
-
-        if (indexWithFileSystemNameCollision != null)
-            throw new IndexCreationException(
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/AbstractIndexCreateController.cs:67]: you are changing the exception

File: src/Raven.Server/ServerWide/Commands/Indexes/PutIndexCommand.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/Commands/Indexes/PutIndexCommand.cs:47]: We rethrow it as `RachisApplyException`.
Changed to the expected exception.

File: src/Raven.Server/Documents/Indexes/AbstractIndexCreateController.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/AbstractIndexCreateController.cs:67]: done

**Pull request merged**